RIOT-OS/RIOT
# drivers/kw2xrf: Improved event callback policy

### Contribution description

kw2xrf driver; kw2xrf_netdev.c:

The event_callback policy was kind of messy; after a TX operation, depending on internal flags, 0, 1 or 2 event callbacks were made, with different parameters.

This change makes sure that a TX operation (T or TR) always results in exactly one event callback.


### Testing procedure
The callback functions currently only increment netstat counters. Making an event callback with the parameter `NETDEV_EVENT_TX_MEDIUM_BUSY` will increment the fail counter. Making an event callback with the parameter `NETDEV_EVENT_TX_COMPLETE` will increment the success counter.
To observe faulty behaviour follow these instructions:

- Compile and flash gnrc_networking example on kw2xrf board
- Type `ifconfig`
- Under "statistics for layer 2" look at the line `TX succeeded <x> errors <y>`

#### Scenario 1: 0 callbacks made

- Disable the NETOPTs `AUTOACK`, `ACK_REQ` and `AUTOCCA`. To do this type these lines into the console:
`ifconfig <interface_id> -autoack`
`ifconfig <interface_id> -ack_req`
`ifconfig <interface_id> -autocca`

- Send a UDP packet to any address
`udp send <address> <port> <data>`
- Type `ifconfig`
- Under "statistics for layer 2" look at the line `TX succeeded <x> errors <y>`; neither x or y will have incremented but we would have expected x to increment by one.

#### Scenario 2: 2 callbacks made

- This scenario happens, when the driver is told by hardware, that a CCA has identified the channel as busy. Place a jammer next to the board. ~~Or, if you don't want to do that, go to `drivers/kw2xrf/kw2xrf_netdev.c` and in line 694 change~~

~~`if (dregs[MKW2XDM_IRQSTS2] & MKW2XDM_IRQSTS2_CCA) {`~~

~~to~~

~~`if ( !( dregs[MKW2XDM_IRQSTS2] & MKW2XDM_IRQSTS2_CCA ) ) {`~~

~~and recompile and flash. This will make the driver interpret a clear channel as a busy channel instead.~~
_Edit: This doesn't quite work, because this only influences the ISR assessments. A CCAIRQ will also terminate the sequence, which does not happen, when using this workaround. Jamming the channel seems to be the only safe way to test this._

- Send a UDP packet to any address
`udp send <address> <port> <data>`
- Type `ifconfig`
- Under "statistics for layer 2" look at the line `TX succeeded <x> errors <y>`; both x and y will have incremented but we only would have expected y to increment by one.


Keep in mind, the board does some stuff in the background too, including receiving and sending data over the radio. This will increment the netstat counters too. To avoid misinterpretations, you may want to enable debug prints.
